### PR TITLE
Hashing logic for c10::complex

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51472 Add support for complex valued keys for dict in TS
* **#51441 Hashing logic for c10::complex**

Differential Revision: [D26170195](https://our.internmc.facebook.com/intern/diff/D26170195)

### CODE REVIEW

File: c10/test/util/complex_test_common.h [hunk 1]
```diff
@@ -167,6 +169,11 @@ TEST(TestConstructors, FromThrust) {
 }
 #endif
 
+TEST(TestConstructors, UnorderedMap) {
+  std::unordered_map<c10::complex<double>, c10::complex<double>> m;
+  m[c10::complex<double>(2.5, 3)] = c10::complex<double>(2, -3.2);
+  m[c10::complex<double>(2, 0)] = c10::complex<double>(0, -3);
```
Comment (ezyang) [inline c10/test/util/complex_test_common.h:175]: This isn't a particularly robust test haha

**Review by ezyang — APPROVED** [review]

File: c10/test/util/complex_test_common.h (hunk 1 above)
Comment (anjali411) [reply c10/test/util/complex_test_common.h:175]: lol yeah I updated the test to access values and check equality

File: c10/test/util/complex_test_common.h [hunk 2]
```diff
@@ -167,6 +169,17 @@ TEST(TestConstructors, FromThrust) {
 }
 #endif
 
+TEST(TestConstructors, UnorderedMap) {
```
Comment (zasdfgbnm) [inline c10/test/util/complex_test_common.h:172]: Why this test is part of `TestConstructors`? Hashing sounds unrelated to "constructor".

Comment (anjali411) [reply c10/test/util/complex_test_common.h:172]: yeah I guess the context in which these tests were named were for testing the construction of `c10::complex`, however it seemed to me that testing construction of containers containing `c10::complex` could fall under this category as well.

**Pull request closed without merging**